sycl/doc/Assert.md

```diff
@@ -0,0 +1,269 @@
+multiple workitems, implementation is required to deliver only one. The
+assertion failure message is printed to `stderr` by SYCL Runtime.
+
+When multiple kernels are enqueued and more than one fail at assertion, at least
+single assertion should be reported.
+
+
+## User requirements
+
+From user's point of view there are the following requirements:
+
+| # | Title | Description | Importance |
+| - | ----- | ----------- | ---------- |
+| 1 | Handle assertion failure | Signal about assertion failure via SYCL asynchronous exception | Must have |
+| 2 | Print assert message | Assert function should print message to stderr at host | Must have |
+| 3 | Stop under debugger | When debugger is attached, break at assertion point | Highly desired |
+| 4 | Reliability | Assert failure should be reported regardless of kernel deadlock | Highly desired |
+
+Implementations without enough capabilities to implement fourth requirement are
+allowed to realize the fallback approach described below, which does not
+guarantee assertion failure delivery to host, but is still useful in many
+practical cases.
+
+## Contents of `sycl::event_error`
+
+Interface of `sycl::event_error` should look like:
+```
+class event_error : public runtime_error {
+public:
+  event_error() = default;
+
+  event_error(const char *Msg, cl_int Err)
+      : event_error(string_class(Msg), Err) {}
+
+  event_error(const string_class &Msg, cl_int Err) : runtime_error(Msg, Err) {}
+};
+```
+
+Regardless of whether asynchronous exception handler is set or not, there's an
+action to be performed by SYCL Runtime. To achieve this, information about
+assert failure should be propagated from device-side to SYCL Runtime. This
+should be performed via calls to `piEventGetInfo`. This Plugin Interface call
+"lowers" to `clGetEventInfo` for OpenCL backend and `zeEventQueryStatus` for
+Level-Zero backend.
+
+
+## Terms
+
+ - Device-side Runtime - runtime library supplied by the Native Device Compiler
+   and running on the device.
+ - Native Device Compiler - compiler which generates device-native binary image
+   based on input SPIR-V image.
+ - Low-level Runtime - the backend/runtime behind DPCPP Runtime attached via the
+   Plugin Interface.
+ - Accessor metadata - parts of accessor representation at device-side: pointer,
+   ranges, offset.
+
+
+## How it works?
+
+`assert(expr)` macro ends up in call to `__devicelib_assert_fail`. This function
+is part of [Device library extension](extensions/C-CXX-StandardLibrary/DeviceLibExtensions.rst#cl_intel_devicelib_cassert).
+
+Implementation of this function is supplied by Native Device Compiler for
+safe approach or by DPCPP Compiler for fallback one.
+
+Due to lack of support of online linking in Level-Zero, the application is
+linked against fallback implementation of `__devicelib_assert_fail`. Hence,
+Native Device Compilers should prefer their implementation instead of the one
+provided in incoming SPIR-V/LLVM IR binary.
```

Comment:
I wonder how is this will be implemented? Are we going to support `weak` linkage type for that? Or `weak` is just implied by the function name? The latter seems a bit hacky - I suggest that we just fix online linking in Level Zero instead

Reply:
I think either explicit weak linking marker for fallback devicelib implementation could work fine for us.

Reply:
Then you have one more dependency: there is no `weak` linkage type available in SPIR-V spec

Reply:
Maybe we can get online linking support in time for this feature?

Reply:
Removed this paragraph.